### PR TITLE
Fix/build on openbsd

### PR DESCRIPTION
**Pull request by knightjoel**
### Issues:
n/a

### Description of changes: 
This PR enables successful builds on OpenBSD. 

### Call-outs:
- Building googletest with `-Wvla` on OpenBSD throws an error. I'm not sure what an appropriate fix is here or if it's even worth it (how critical is getting the thread count?). As a starting point, I've just removed the offending code.

### Testing:
Test builds done on OpenBSD 7.2/amd64. Also verified build is still successful on FreeBSD 13/amd64.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and the ISC license.

### CODE REVIEW

File: include/openssl/thread.h [hunk 1]
```diff
@@ -59,6 +59,10 @@
 
 #include <sys/types.h>
 
+#if defined(__OpenBSD__)
+#   include <pthread.h>
+#endif
```
Comment (PiotrSikora) [inline include/openssl/thread.h:64]: Nit: This should be moved inside the `!defined(__GLIBC__)` block to respect `OPENSSL_THREADS`.

File: include/openssl/thread.h [hunk 2]
```diff
@@ -59,6 +59,10 @@
 
 #include <sys/types.h>
 
+#if defined(__OpenBSD__)
```
Comment (PiotrSikora) [inline include/openssl/thread.h:62]: Nit: This should be moved to `include/openssl/base.h` and defined as `OPENSSL_OPENBSD`, which should be used here instead. This is how it's done for other operating systems (e.g. FreeBSD).

**Comment by torben-hansen** [comment]
What's the entire log output for?
```
unknown type name 'pthread_rwlock_t
```

**Comment by knightjoel** [comment]
> What's the entire log output for?
> 
> ```
> unknown type name 'pthread_rwlock_t
> ```

```
In file included from /home/joel/tmp/aws-lc/crypto/fipsmodule/cpucap/cpucap.c:12:
In file included from /home/joel/tmp/aws-lc/crypto/../include/openssl/cpu.h:18:
In file included from /home/joel/tmp/aws-lc/crypto/../include/openssl/crypto.h:27:
/home/joel/tmp/aws-lc/crypto/../include/openssl/thread.h:81:9: error: unknown type name 'pthread_rwlock_t'
typedef pthread_rwlock_t CRYPTO_MUTEX;
        ^
1 error generated.
```

**Comment by torben-hansen** [comment]
I not sure `GetThreadCount()` is that valuable, it checks whether there is actually more threads. Newer versions don't have this functions I think.

File: include/openssl/thread.h (hunk 1 above)
Comment (torben-hansen) [reply include/openssl/thread.h:64]: Agree.

This leaks `<pthread.h>` everywhere since `thread.h` is a public header... Not sure I like that...

Comment (knightjoel) [reply include/openssl/thread.h]: Oof, I hadn't considered that. Thanks for the pointer. How does the updated diff look to you?

File: include/openssl/thread.h (hunk 2 above)
Comment (knightjoel) [reply include/openssl/thread.h]: Updated the diff. Please let me know how close/far I am to what you were expecting.

**Review by PiotrSikora — COMMENTED** [review]
Updated diff LGTM for the `include/`.

As for `third_party/googletest`, it looks that it wasn't updated in 2 years, perhaps it's easier to update it than patch locally (assuming it's still used in AWS-LC)?

**Comment by knightjoel** [comment]
> Updated diff LGTM for the `include/`.

Thanks 👍 
 
> As for `third_party/googletest`, it looks that it wasn't updated in 2 years, perhaps it's easier to update it than patch locally (assuming it's still used in AWS-LC)?

Upstream, that bit of code is the same as what's in aws-lc. Upstream doesn't compile with `-Wvla` and so it's unaffected. I considered sending a patch upstream, but they don't have much incentive to fix this given it doesn't directly affect them.

**Review by torben-hansen — DISMISSED** [review]

**Comment by torben-hansen** [comment]
Note that there is no CI dimension for OpenBSD, so this is **experimental**.

**Comment by knightjoel** [comment]
> Note that there is no CI dimension for OpenBSD, so this is **experimental**.

Would having CI for OpenBSD be of interest? It would have to run as a GitHub action given there is no support for OpenBSD in CodeBuild.

**Review by torben-hansen — APPROVED** [review]

**Review by andrewhop — APPROVED** [review]

**Comment by knightjoel** [comment]
Hi folks, would someone be able to merge this? Please and thank you.

**Pull request merged**